src/middlewared/middlewared/plugins/auth.py

```diff
@@ -409,103 +402,494 @@ async def login(self, app, username, password, otp_token):
+                                'token': token.token,
+                            }
+                        },
+                        'error': 'Bad token',
+                    }, False)
+                    return response
+
+                cred = TokenSessionManagerCredentials(self.token_manager, token)
+                await login_fn(app, cred)
+                resp = {
+                    'pam_response': {
+                        'code': pam.PAM_SUCCESS,
+                        'reason': None
+                    }
                 }
-            },
-            "error": "Invalid API key",
-        }, False)
-        return False
+
+            case _:
+                # This shouldn't happen so we'll log it and raise a call error
+                self.logger.error('%s: unexpected authentication mechanism', mechanism)
+                raise CallError(f'{mechanism}: unexpected authentication mechanism')
+
+        match resp['pam_response']['code']:
+            case pam.PAM_SUCCESS:
+                response['response_type'] = AuthResp.SUCCESS
+                if data['login_options']['user_info']:
+                    response['user_info'] = await self.me(app)
+                else:
+                    response['user_info'] = None
+
+            case pam.PAM_AUTH_ERR | pam.PAM_USER_UNKNOWN:
+                # We have to squash AUTH_ERR and USER_UNKNOWN into a generic response
+                # to prevent unauthenticated remote clients from guessing valid usernames.
+                response['response_type'] = AuthResp.AUTH_ERR
+            case pam.PAM_ACCT_EXPIRED | pam.PAM_NEW_AUTHTOK_REQD | pam.PAM_CRED_EXPIRED:
+                response['response_type'] = AuthResp.EXPIRED.name
+            case _:
+                # This is unexpected and so we should generate a debug message
+                # so that we can better handle in the future.
+                self.logger.debug(
+                    '%s: unexpected response code [%d] to authentication request',
+                    mechanism, resp['pam_response']['code']
+                )
+                response['response_type'] = AuthResp.AUTH_ERR
+
+        return response
+
+    @private
+    @pass_app()
+    async def get_login_user(self, app, username, password, mechanism):
+        otp_required = False
+        resp = await self.middleware.call(
+            'auth.authenticate_plain',
+            username, password,
+            mechanism == AuthMech.API_KEY_PLAIN,
+            app=app
+        )
+        if mechanism == AuthMech.PASSWORD_PLAIN and resp['pam_response']['code'] == pam.PAM_SUCCESS:
+            twofactor_auth = await self.middleware.call('auth.twofactor.config')
+            if twofactor_auth['enabled'] and '2FA' in resp['user_data']['account_attributes']:
+                otp_required = True
+
+        return resp | {'otp_required': otp_required}
 
     @cli_private
     @no_auth_required
-    @accepts(Password('token'))
+    @accepts(Password('api_key'))
```

Comment:
While you're here, we should change the old `@accepts` decorator to the new one (also get rid of `@returns`